### PR TITLE
Remove use of rules batching for upmap balancer and default for upmap_max_deviation to 5

### PR DESCRIPTION
**Pull request by dzafman**
## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dzafman** [comment]
@neha-ojha Re-testing performance of rules based without Debug enabled before merging this.

File: src/tools/osdmaptool.cc [hunk 1]
```diff
@@ -403,14 +408,24 @@ int main(int argc, const char **argv)
       cout << "No pools available" << std::endl;
       goto skip_upmap;
     }
+    int rounds = 0;
+    struct timespec round_start;
+    int r = clock_gettime(CLOCK_MONOTONIC, &round_start);
+    assert(r == 0);
+    do {
```
Comment (dzafman) [inline src/tools/osdmaptool.cc]: I'm indenting the code here.

**Comment by xiexingguo** [comment]
IIRC, https://github.com/ceph/ceph/pull/32247/commits/9b21f1e6888aed35cedc4661736304150369ba90 aims to fix a bug that two or more pools that share some common osds could possible race with each when balancer module is enabled.

I am starting working on a series of patches to improve the efficiency of balancer, e.g., :

https://github.com/ceph/ceph/pull/32266
https://github.com/ceph/ceph/pull/32289

IMO in the long term we don't have to revert the above commit...

**Comment by dzafman** [comment]
> 
> IMO in the long term we don't have to revert the above commit...

@xiexingguo 

Interestingly, using some real world contributed osdmaps, the different was fractions of a second sometimes better sometimes worse. However, there are large cluster cases with rules batching that hurts performance.  

These example are using the new code with upmap_max_deviation 5 and upmap_max_optimizations 10 and using the --upmap-active option of the osdmaptool to simulate an active balancer:

`osdmaptool --upmap upmap.out --upmap-active --upmap-deviation 5 --upmap-max 10 osdmap`

Example 1 (includes 3 - 9 shard EC pool with 64, 1024 and 2048 PGs with the same rule)

Description        |By pool         | Rules Batched
------------ |------------ | ---------------
Rounds of 10 upmaps | 80 rounds  |  112 rounds
Max secs per round|2.09631 |   110.086

Example 2 (includes 2 - 11 shard EC pool with 8k PGs each with the same rule)

Description        |By pool         | Rules Batched
------------ |------------ | ---------------
Rounds of 10 upmaps | 353 rounds  |  430 rounds
Max secs per round| 9.68959 |   54.7348*

* The compute time doesn't tell us how long calc_pg_upmaps() ran for an individual pool or rule batch it does reflect the maximum time to create a "plan" or one cycle in active mode.  54 seconds is a long time for something that is supposed to run once per minute.

**Comment by dzafman** [comment]
@neha-ojha More documentation updates coming.

File: doc/man/8/osdmaptool.rst [hunk 2]
```diff
@@ -131,6 +132,34 @@ Options
...
+
+   set max upmap entries to calculate [default: 10]
+
+.. option:: --upmap-deviation <max-deviation>
+
+   max deviation from target [default: 5]
+
+.. option:: --upmap-pool <poolname>
+
+   restrict upmap balancing to 1 or more pools
```
Comment (jdurgin) [inline doc/man/8/osdmaptool.rst]: the syntax shown only looks like one pool

File: src/tools/osdmaptool.cc (hunk 1 above)
Comment (jdurgin) [reply src/tools/osdmaptool.cc]: yes, or move the body to a helper function

File: doc/man/8/osdmaptool.rst [hunk 3]
```diff
@@ -21,6 +27,8 @@ Description
 **osdmaptool** is a utility that lets you create, view, and manipulate
 OSD cluster maps from the Ceph distributed storage system. Notably, it
 lets you extract the embedded CRUSH map or import a new CRUSH map.
+It can also simulate the upmap balancer mode so you can get a sense of
+what is needed to balancer your PGs.
```
Comment (neha-ojha) [inline doc/man/8/osdmaptool.rst]:
```suggestion
what is needed to balance your PGs.
```

File: doc/man/8/osdmaptool.rst [hunk 4]
```diff
@@ -119,6 +131,43 @@ Options
...
+
+   restrict upmap balancing to 1 pool or the option can be repeated for multiple pools
+
+.. option:: --upmap-save
+
+   write modified OSDMap with upmap changes
+
+.. option:: --upmap-active
+
+   Act like an active balancer keep applying changes until balanced
```
Comment (neha-ojha) [inline doc/man/8/osdmaptool.rst]:
```suggestion
   Act like an active balancer, keep applying changes until balanced
```

**Comment by dzafman** [comment]
I'm rerunning the balancer test.

**Comment by dzafman** [comment]
Testing passed (balancer.sh test)
http://pulpito.ceph.com/dzafman-2020-01-06_21:36:46-rados-master-distro-basic-smithi/4644116/

**Pull request merged**